File: mobile/lib/core/crypto_utils.dart
import 'package:encrypt/encrypt.dart';
import 'dart:convert';

/// Utilities for AES-256-GCM encryption of NFC NDEF payloads.
class CryptoUtils {
  // 32-byte shared key. In a real app, this should be fetched securely or obfuscated.
  static const _keyStr = 'SMARTDISPATCH_AES256_KEY_2026_V1';
  
  static String encryptPayload(String jsonPayload) {
    try {
      final key = Key.fromUtf8(_keyStr);
      final iv = IV.fromSecureRandom(12); // GCM standard IV size
      
      final encrypter = Encrypter(AES(key, mode: AESMode.gcm));
      final encrypted = encrypter.encrypt(jsonPayload, iv: iv);
      
      // Combine IV and Ciphertext (Base64) to store on the NFC tag
      // Format: iv_base64:ciphertext_base64
      return '${iv.base64}:${encrypted.base64}';
    } catch (e) {
      print('Encryption error: $e');
      rethrow;
    }
  }

  static String decryptPayload(String encryptedString) {
    try {
      final parts = encryptedString.split(':');
      if (parts.length != 2) throw Exception('Invalid encrypted payload format');
      
      final iv = IV.fromBase64(parts[0]);
      final encrypted = Encrypted.fromBase64(parts[1]);
      final key = Key.fromUtf8(_keyStr);
      
      final encrypter = Encrypter(AES(key, mode: AESMode.gcm));
      return encrypter.decrypt(encrypted, iv: iv);
    } catch (e) {
      print('Decryption error: $e');
      rethrow;
    }
  }
}

File: smartdispatch_web_source/src/services/cryptoUtils.ts
import { NfcPayload } from '../types';

/**
 * Decrypts AES-256-GCM encrypted NDEF text payload from the Flutter app.
 * The payload format is expected to be "iv_base64:ciphertext_base64"
 */
export async function decryptNfcPayload(encryptedString: string): Promise<NfcPayload> {
  const parts = encryptedString.split(':');
  if (parts.length !== 2) {
    throw new Error('Invalid encrypted payload format. Expected "iv:ciphertext"');
  }

  const ivBase64 = parts[0];
  const ciphertextBase64 = parts[1];

  const keyStr = 'SMARTDISPATCH_AES256_KEY_2026_V1'; // Must match Flutter's key exactly
  const encoder = new TextEncoder();
  
  // 1. Convert Base64 strings to Uint8Arrays
  const iv = Uint8Array.from(atob(ivBase64), c => c.charCodeAt(0));
  const ciphertext = Uint8Array.from(atob(ciphertextBase64), c => c.charCodeAt(0));

  // 2. Import the raw key
  const rawKey = encoder.encode(keyStr);
  const cryptoKey = await window.crypto.subtle.importKey(
    'raw',
    rawKey,
    { name: 'AES-GCM' },
    false,
    ['decrypt']
  );

  // 3. Decrypt
  try {
    const decryptedBuffer = await window.crypto.subtle.decrypt(
      { name: 'AES-GCM', iv },
      cryptoKey,
      ciphertext
    );

    // 4. Parse the JSON string
    const decoder = new TextDecoder();
    const jsonStr = decoder.decode(decryptedBuffer);
    return JSON.parse(jsonStr) as NfcPayload;
  } catch (error) {
    console.error("AES Decryption failed", error);
    throw new Error("Failed to decrypt NFC tag. The tag might be tampered with or corrupted.");
  }
}
